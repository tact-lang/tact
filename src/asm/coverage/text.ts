import { CoverageSummary, InstructionStat, Line } from "@/asm/coverage/data";
import { calculateTotalGas } from "@/asm/coverage/html";

export const generateShortSummary = (summary: CoverageSummary): string => {
    return [
        "Coverage Summary:",
        `Lines: ${summary.coveredLines}/${summary.totalLines} (${summary.coveragePercentage.toFixed(2)}%)`,
        `Total Gas: ${summary.totalGas}`,
        `Total Hits: ${summary.totalHits}`,
    ].join("\n");
};

export const generateTextReport = (
    lines: readonly Line[],
    summary: CoverageSummary,
): string => {
    const maxLineNumberWidth = lines.length.toString().length;

    const annotatedLines = lines
        .map((line, index) => {
            const { gasInfo, hitsInfo, status } = lineInfo(line);

            const lineNumber = index + 1;
            const lineNumberPres = lineNumber
                .toString()
                .padStart(maxLineNumberWidth);

            return `${lineNumberPres} ${status}| ${line.line.padEnd(40)} |${gasInfo.padEnd(10)} |${hitsInfo}`;
        })
        .join("\n");

    const summaryText = [
        "Coverage Summary:",
        `Lines: ${summary.coveredLines}/${summary.totalLines} (${summary.coveragePercentage.toFixed(2)}%)`,
        `Total Gas: ${summary.totalGas}`,
        `Total Hits: ${summary.totalHits}`,
        "",
        "Instruction Stats:",
        ...instructionsStats(summary),
    ].join("\n");

    return `${summaryText}\n\nAnnotated Code:\n${annotatedLines}`;
};

type LineInfo = {
    readonly gasInfo: string;
    readonly hitsInfo: string;
    readonly status: string;
};

const lineInfo = (line: Line): LineInfo => {
    if (line.info.$ === "Covered") {
        const totalGas = calculateTotalGas(line.info.gasCosts);
        const gasInfo = ` gas:${totalGas}`;
        const hitInfo = ` hits:${line.info.hits}`;
        return { gasInfo, hitsInfo: hitInfo, status: "✓ " };
    }

    if (line.info.$ === "Uncovered") {
        return { gasInfo: "", hitsInfo: "", status: "✗ " };
    }

    return { gasInfo: "", hitsInfo: "", status: "  " };
};

const instructionsStats = (summary: CoverageSummary) =>
    summary.instructionStats.map((stat) =>
        formatInstructionStat(stat, summary.totalGas),
    );

const formatInstructionStat = (stat: InstructionStat, totalGas: number) => {
    const name = stat.name.padEnd(15);
    const totalGasStr = stat.totalGas.toString().padEnd(3, "");
    const percent = ((stat.totalGas / totalGas) * 100).toFixed(2);
    return `  ${name} | ${totalGasStr} gas | ${stat.totalHits} hits | ${stat.avgGas} avg gas | ${percent}%`;
};
