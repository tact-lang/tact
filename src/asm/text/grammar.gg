File = instructions:instructions;

instructions = (Instruction / ExplicitRef / EmbedSlice / Exotic)*;

Instruction = name:Id args:Argument*;

Argument = expression:(
      IntegerLiteral
    / DataLiteral
    / Code
    / Dictionary
    / StackElement
    / ControlRegister
);

StackElement = value:$#("s" "-"? IntegerLiteralDec);
ControlRegister = value:$#("c" IntegerLiteralDec);

Code = "{" instructions:instructions "}";

Dictionary = "[" entries:DictionaryEntry* "]";

DictionaryEntry = id:IntegerLiteral "=>" code:Code;

ExplicitRef = keyword<"ref"> code:Code;
EmbedSlice  = keyword<"embed"> data:DataLiteral;
Exotic      = keyword<"exotic"> lib:(ExoticLibrary / DefaultExotic);

ExoticLibrary = keyword<"library"> data:DataLiteral;
DefaultExotic = data:DataLiteral;

idPart "identifier character" = [a-zA-Z0-9_];
Id "identifier" = name:#$(!reservedWord [a-zA-Z_] idPart*);

DataLiteral = value:(HexLiteral / BinLiteral / BocLiteral / StringLiteral);

HexLiteral = "x{" content:$(hexDigit* "_"?) "}";
BinLiteral = "b{" content:$(binDigit*) "}";
BocLiteral = "boc{" content:$(hexDigit*) "}";

// Literals
IntegerLiteral = op:"-"? value:(IntegerLiteralHex / IntegerLiteralBin / IntegerLiteralOct / IntegerLiteralDec);

IntegerLiteralDec = digits:#underscored<digit>;
IntegerLiteralHex = digits:#("0" [xX] @underscored<hexDigit>);
IntegerLiteralBin = digits:#("0" [bB] @underscored<binDigit>);
IntegerLiteralOct = digits:#("0" [oO] @underscored<[0-7]>);

underscored<T> = $(T ("_"? T)*);

hexDigit "hexadecimal digit" = [0-9a-fA-F];
binDigit "binary digit" = [01];
digit "digit" = [0-9];

StringLiteral = value:#("\"" @$([^"\\])* "\"");

keyword<T> = #(@T !idPart);

reservedWord "reserved word" = keyword<(
    "ref" / "embed" / "exotic" / "library"
)>;

space "space" = [ \t\r\n] / comment;
comment = singleLineComment;
singleLineComment = "//" @$[^\r\n]*;
