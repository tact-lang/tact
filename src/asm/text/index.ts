export * from "./convert"
export * as ast from "./grammar"
export * from "./parse"
export * from "./printer"
export * from "./printer-gen"
export * from "./util"
