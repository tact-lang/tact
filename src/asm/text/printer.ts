import type { Instr } from "@/asm/runtime";
import type { Code, Dict } from "@/asm/runtime/util";
import type { Slice } from "@ton/core";
import { printInstruction } from "@/asm/text/printer-gen";

export class Printer {
    public indent: number;
    public result: string;

    public constructor(indent: number) {
        this.indent = indent;
        this.result = "";
    }

    public beginLine(text: string) {
        const indentText = "    ".repeat(this.indent);
        this.result += indentText + text;
    }

    public append(text: string) {
        this.result += text;
    }

    public inIndent(action: (p: Printer) => void) {
        this.indent++;
        action(this);
        this.indent--;
    }

    public build(): string {
        return this.result;
    }
}

export const printInstr = (p: Printer, instr: Instr) => {
    if (instr.$ === "PSEUDO_PUSHREF") {
        p.beginLine("ref ");
        printCode(p, instr.arg0);
        return;
    }

    if (instr.$ === "PSEUDO_PUSHSLICE") {
        p.beginLine("embed ");
        printSlice(p, instr.arg0);
        return;
    }

    if (instr.$ === "PSEUDO_EXOTIC") {
        p.beginLine("exotic ");

        if (instr.arg0.$ === "LibraryCell") {
            p.append("library ");
            printSlice(p, instr.arg0.data);
            return;
        }

        printSlice(p, instr.arg0.cell.beginParse(true));
        return;
    }

    printInstruction(p, instr);
};

export const printInstructions = (p: Printer, instructions: Instr[]) => {
    if (instructions.length === 0) {
        p.append("{}");
        return;
    }

    p.append("{\n");
    p.inIndent((p) => {
        for (const instr of instructions) {
            printInstr(p, instr);
            p.append("\n");
        }
    });
    p.beginLine("}");
};

export const print = (instructions: Instr[]) => {
    const p = new Printer(0);
    for (const instr of instructions) {
        printInstr(p, instr);
        p.append("\n");
    }
    return p.build();
};

export function printDictionary(p: Printer, dict: Dict) {
    if (dict.$ === "RawDict") {
        printSlice(p, dict.slice);
    }

    if (dict.$ === "DecompiledDict") {
        p.append(`[\n`);

        for (const method of dict.methods) {
            p.inIndent((p) => {
                p.beginLine(`${method.id} => `);
                printInstructions(p, method.instructions);
                p.append(`\n`);
            });
        }

        p.beginLine("]");
    }
}

export function printCode(p: Printer, arg0: Code) {
    if (arg0.$ === "Instructions") {
        printInstructions(p, arg0.instructions);
    }

    if (arg0.$ === "Raw") {
        const value = arg0.slice.asCell().toString().slice(2, -1);
        if (value.includes("\n")) {
            // cell with refs
            p.append(`boc{${arg0.slice.asCell().toBoc().toString("hex")}}`);
        } else {
            p.append(`x{${value}}`);
        }
    }
}

export function printSlice(p: Printer, slice: Slice) {
    const value = slice.asCell().toString().slice(2, -1);
    if (value.includes("\n")) {
        // cell with refs
        p.append(`boc{${slice.asCell().toBoc().toString("hex")}}`);
    } else {
        p.append(`x{${value}}`);
    }
}
