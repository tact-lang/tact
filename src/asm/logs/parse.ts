import * as $ from "@tonstudio/parser-runtime";
import * as G from "@/asm/logs/grammar";
import type { Stack, StackElement } from "@/asm/logs/stack";
import VmParsedStack = G.$ast.VmParsedStack;

export type VmLine =
    | VmLoc
    | VmStack
    | VmExecute
    | VmLimitChanged
    | VmGasRemaining
    | VmException
    | VmExceptionHandler
    | VmFinalC5
    | VmUnknown;

export type VmLoc = {
    readonly $: "VmLoc";
    readonly hash: string;
    readonly offset: number;
};

export type VmStack = {
    readonly $: "VmStack";
    readonly stack: Stack;
};

export type VmExecute = {
    readonly $: "VmExecute";
    readonly instr: string;
};

export type VmLimitChanged = {
    readonly $: "VmLimitChanged";
    readonly limit: number;
};

export type VmGasRemaining = {
    readonly $: "VmGasRemaining";
    readonly gas: number;
};

export type VmException = {
    readonly $: "VmException";
    readonly errno: number;
    readonly message: string;
};

export type VmExceptionHandler = {
    readonly $: "VmExceptionHandler";
    readonly errno: number;
};

export type VmFinalC5 = {
    readonly $: "VmFinalC5";
    readonly hex: string;
};

export type VmUnknown = {
    readonly $: "VmUnknown";
    readonly text: string;
};

export const parse = (log: string): VmLine[] => {
    const lines = log.split("\n");
    const vmLines = lines.map((line) => parseLine(line));

    return vmLines.map((it) => processVmLine(it));
};

const parseLine = (line: string): G.$ast.vmLine => {
    const res = $.parse({
        grammar: G.vmLine,
        space: G.space,
        text: line,
    });
    if (res.$ === "success") {
        return res.value;
    }
    return {
        $: "VmUnknown",
        text: line,
        loc: $.emptyLoc(0),
    };
};

const processVmLine = (line: G.$ast.vmLine): VmLine => {
    switch (line.$) {
        case "VmLoc":
            return {
                $: "VmLoc",
                hash: line.hash.trim(),
                offset: parseNumber(line.offset),
            };
        case "VmStack": {
            const stack = parseStack(line.stack);
            if (!stack) {
                throw new Error(`Cannot parse stack: ${line.stack}`);
            }

            return {
                $: "VmStack",
                stack: processStack(stack),
            };
        }
        case "VmExecute":
            return {
                $: "VmExecute",
                instr: line.instr.trim(),
            };
        case "VmLimitChanged":
            return {
                $: "VmLimitChanged",
                limit: parseNumber(line.limit),
            };
        case "VmGasRemaining":
            return {
                $: "VmGasRemaining",
                gas: parseNumber(line.gas),
            };
        case "VmException":
            return {
                $: "VmException",
                errno: parseNumber(line.errno),
                message: line.message.trim(),
            };
        case "VmExceptionHandler":
            return {
                $: "VmExceptionHandler",
                errno: parseNumber(line.errno),
            };
        case "VmFinalC5":
            return {
                $: "VmFinalC5",
                hex: line.value.value.trim(),
            };
        case "VmUnknown":
            return {
                $: "VmUnknown",
                text: line.text.trim(),
            };
        default:
            return {
                $: "VmUnknown",
                text: "",
            };
    }
};

const parseStack = (line: string) => {
    const res = tryParseStack(line);
    if (res) return res;

    const res2 = tryParseStack(line + "]");
    if (res2) return res2;

    const res3 = tryParseStack(line + "} ]");
    if (res3) return res3;

    // try to recover many missing `]`
    for (let i = 0; i < 100; i++) {
        const resN = tryParseStack(line + "} ]" + "]".repeat(i));
        if (resN) {
            return resN;
        }
    }

    return undefined;
};

const tryParseStack = (line: string) => {
    const res = $.parse({
        grammar: G.VmParsedStack,
        space: G.space,
        text: line,
    });
    if (res.$ === "success") {
        return res.value;
    }
    return undefined;
};

const processStack = (stack: VmParsedStack): StackElement[] => {
    return stack.values.map((it) => processStackElement(it));
};

const parseNumber = (it: G.$ast.$number) => {
    const val = Number.parseInt(it.value);
    if (it.op === "-") {
        return -val;
    }
    return val;
};

const parseBigNum = (it: G.$ast.$number) => {
    const val = BigInt(it.value);
    if (it.op === "-") {
        return -val;
    }
    return val;
};

const processStackElement = (it: G.$ast.VmStackValue): StackElement => {
    switch (it.value.$) {
        case "Null":
            return { $: "Null" };
        case "NaN":
            return { $: "NaN" };
        case "Integer":
            return { $: "Integer", value: parseBigNum(it.value.value) };
        case "Tuple":
        case "TupleParen":
            return {
                $: "Tuple",
                elements: it.value.elements.map((it) =>
                    processStackElement(it),
                ),
            };
        case "Cell":
            return {
                $: "Cell",
                boc: it.value.value,
            };
        case "Continuation":
            return {
                $: "Continuation",
                name: it.value.value,
            };
        case "Builder":
            return {
                $: "Builder",
                hex: it.value.value,
            };
        case "CellSlice":
            if (it.value.body.$ === "CellSliceBody") {
                return {
                    $: "Slice",
                    hex: it.value.body.value,
                    startBit: parseNumber(it.value.body.bits.start),
                    endBit: parseNumber(it.value.body.bits.end),
                    startRef: parseNumber(it.value.body.refs.start),
                    endRef: parseNumber(it.value.body.refs.end),
                };
            }

            return {
                $: "Slice",
                hex: it.value.body.value,
                startBit: 0,
                endBit: 0,
                startRef: 0,
                endRef: 0,
            };
        case "Unknown":
            return {
                $: "Unknown",
                value: "",
            };
        default:
            return { $: "Unknown", value: "" };
    }
};
