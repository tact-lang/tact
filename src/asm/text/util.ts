import type * as $ from "@tonstudio/parser-runtime";
import type { Code, DecompiledMethod, Dict, Loc } from "@/asm/runtime/util";
import {
    boc,
    decompiledCode,
    DefaultExoticCell,
    hex,
    bin,
    LibraryCell,
    rawCode,
    code,
} from "@/asm/runtime/util";
import type { $ast } from "@/asm/text/grammar";
import * as i from "@/asm/runtime";
import { convertInstruction } from "@/asm/text/convert";
import type { Slice } from "@ton/core";
import { beginCell } from "@ton/core";

export type Ctx = {
    readonly lines: readonly string[];
    readonly filepath: string;
};

export type Convert = (ctx: Ctx, instr: $ast.Instruction, loc: Loc) => i.Instr;

export class ParseError extends Error {
    public loc: Loc;
    public msg: string;

    public constructor(loc: Loc, msg: string) {
        super(msg);
        this.name = "ParseError";
        this.loc = loc;
        this.msg = msg;
    }

    public override toString() {
        return `${this.name}: ${this.msg} at ${this.loc.file}:${this.loc.line}`;
    }
}

const offsetToLine = (
    lines: readonly string[],
    searchOffset: number,
): number => {
    let offset = 0;
    let index = 0;
    for (const line of lines) {
        offset += line.length + 1;
        if (searchOffset < offset) {
            return index;
        }
        index++;
    }

    return lines.length - 1;
};

export const createLoc = (ctx: Ctx, loc: $.Loc): Loc => {
    if (loc.$ === "empty") {
        return {
            file: ctx.filepath,
            line: offsetToLine(ctx.lines, loc.at),
        };
    }

    return {
        file: ctx.filepath,
        line: offsetToLine(ctx.lines, loc.start),
    };
};

export const processInstructions = (
    ctx: Ctx,
    instructions: $ast.instructions,
): i.Instr[] => {
    return instructions.map((it) => {
        const loc = createLoc(ctx, it.loc);
        try {
            if (it.$ === "ExplicitRef") {
                return i.PSEUDO_PUSHREF(
                    i.util.code(processInstructions(ctx, it.code.instructions)),
                    loc,
                );
            }

            if (it.$ === "EmbedSlice") {
                return i.PSEUDO_PUSHSLICE(parseDataLiteral(it.data), loc);
            }

            if (it.$ === "Exotic") {
                if (it.lib.$ === "DefaultExotic") {
                    return i.PSEUDO_EXOTIC(
                        DefaultExoticCell(
                            parseDataLiteral(it.lib.data).asCell(),
                        ),
                    );
                }

                return i.PSEUDO_EXOTIC(
                    LibraryCell(parseDataLiteral(it.lib.data)),
                    loc,
                );
            }

            return convertInstruction(ctx, it, loc);
        } catch (error) {
            if (typeof error === "string") {
                throw new ParseError(loc, error);
            }

            throw error;
        }
    });
};

export const singleIntegerArg = (instr: $ast.Instruction) => {
    const [arg0raw] = instr.args;
    if (!arg0raw) {
        throw new Error(`Expected 1 argument`);
    }

    const arg = arg0raw.expression;
    if (arg.$ !== "IntegerLiteral") {
        throw new Error(`Expected integer literal argument, got ${arg.$}`);
    }

    return parseNumber(arg);
};

export const singleStackArg = (instr: $ast.Instruction) => {
    const [arg0raw] = instr.args;
    if (!arg0raw) {
        throw new Error(`Expected 1 argument`);
    }

    const arg = arg0raw.expression;
    if (arg.$ !== "StackElement") {
        throw new Error(`Expected stack element argument, got ${arg.$}`);
    }

    return parseStackElement(arg);
};

export const singleControlArg = (instr: $ast.Instruction) => {
    const [arg0raw] = instr.args;
    if (!arg0raw) {
        throw new Error(`Expected 1 argument`);
    }

    const arg = arg0raw.expression;
    if (arg.$ !== "ControlRegister") {
        throw new Error(`Expected control register argument, got ${arg.$}`);
    }

    const number = Number.parseInt(arg.value.slice(1));
    if (number === 6) {
        throw new Error(`c6 doesn't exist`);
    }

    return number;
};

export const singleBigIntArg = (instr: $ast.Instruction) => {
    const [arg0raw] = instr.args;
    if (!arg0raw) {
        throw new Error(`Expected 1 argument`);
    }

    const arg = arg0raw.expression;
    if (arg.$ !== "IntegerLiteral") {
        throw new Error(`Expected integer literal argument, got ${arg.$}`);
    }

    const value = BigInt(arg.value.digits);
    if (arg.op === "-") {
        return -value;
    }
    return value;
};

export const twoIntegerArgs = (instr: $ast.Instruction): [number, number] => {
    const [arg0raw, arg1raw] = instr.args;
    if (!arg0raw || !arg1raw) {
        throw new Error(`Expected 2 arguments`);
    }

    const arg = arg0raw.expression;
    if (arg.$ !== "IntegerLiteral") {
        throw new Error(`Expected integer literal argument 1, got ${arg.$}`);
    }

    const arg2 = arg1raw.expression;
    if (arg2.$ !== "IntegerLiteral") {
        throw new Error(`Expected integer literal argument 2, got ${arg2.$}`);
    }

    return [parseNumber(arg), parseNumber(arg2)];
};

export const twoStackArgs = (instr: $ast.Instruction): [number, number] => {
    const [arg0raw, arg1raw] = instr.args;
    if (!arg0raw || !arg1raw) {
        throw new Error(`Expected 2 arguments`);
    }

    const arg = arg0raw.expression;
    if (arg.$ !== "StackElement") {
        throw new Error(`Expected stack element argument 1, got ${arg.$}`);
    }

    const arg2 = arg1raw.expression;
    if (arg2.$ !== "StackElement") {
        throw new Error(`Expected stack element argument 2, got ${arg2.$}`);
    }

    return [parseStackElement(arg), parseStackElement(arg2)];
};

export const threeIntegerArgs = (
    instr: $ast.Instruction,
): [number, number, number] => {
    const [arg0raw, arg1raw, arg2raw] = instr.args;
    if (!arg0raw || !arg1raw || !arg2raw) {
        throw new Error(`Expected 3 arguments`);
    }

    const arg = arg0raw.expression;
    if (arg.$ !== "IntegerLiteral") {
        throw new Error(`Expected integer literal argument 1, got ${arg.$}`);
    }

    const arg2 = arg1raw.expression;
    if (arg2.$ !== "IntegerLiteral") {
        throw new Error(`Expected integer literal argument 2, got ${arg2.$}`);
    }

    const arg3 = arg2raw.expression;
    if (arg3.$ !== "IntegerLiteral") {
        throw new Error(`Expected integer literal argument 3, got ${arg3.$}`);
    }

    return [parseNumber(arg), parseNumber(arg2), parseNumber(arg3)];
};

export const threeStackArgs = (
    instr: $ast.Instruction,
): [number, number, number] => {
    const [arg0raw, arg1raw, arg2raw] = instr.args;
    if (!arg0raw || !arg1raw || !arg2raw) {
        throw new Error(`Expected 3 arguments`);
    }

    const arg = arg0raw.expression;
    if (arg.$ !== "StackElement") {
        throw new Error(`Expected stack element argument 1, got ${arg.$}`);
    }

    const arg2 = arg1raw.expression;
    if (arg2.$ !== "StackElement") {
        throw new Error(`Expected stack element argument 2, got ${arg2.$}`);
    }

    const arg3 = arg2raw.expression;
    if (arg3.$ !== "StackElement") {
        throw new Error(`Expected stack element argument 3, got ${arg3.$}`);
    }

    return [
        parseStackElement(arg),
        parseStackElement(arg2),
        parseStackElement(arg3),
    ];
};

export const codeSliceArg = (ctx: Ctx, instr: $ast.Instruction) => {
    const [arg0raw] = instr.args;
    if (!arg0raw) {
        throw new Error(`Expected 1 argument`);
    }

    const arg = arg0raw.expression;
    if (arg.$ !== "Code" && arg.$ !== "DataLiteral") {
        throw new Error(`Expected code or data literal argument, got ${arg.$}`);
    }

    if (arg.$ === "DataLiteral") {
        return processRawSliceCode(arg);
    }

    return decompiledCode(processInstructions(ctx, arg.instructions));
};

export const twoCodeSliceArgs = (
    ctx: Ctx,
    instr: $ast.Instruction,
): [Code, Code] => {
    const [arg0raw, arg1raw] = instr.args;
    if (!arg0raw || !arg1raw) {
        throw new Error(`Expected 2 arguments`);
    }

    const arg = arg0raw.expression;
    if (arg.$ !== "Code" && arg.$ !== "DataLiteral") {
        throw new Error(
            `Expected code or data literal argument 1, got ${arg.$}`,
        );
    }

    const arg1 = arg1raw.expression;
    if (arg1.$ !== "Code" && arg1.$ !== "DataLiteral") {
        throw new Error(
            `Expected code or data literal argument 2, got ${arg1.$}`,
        );
    }

    const first: Code =
        arg.$ === "DataLiteral"
            ? processRawSliceCode(arg)
            : decompiledCode(processInstructions(ctx, arg.instructions));
    const second =
        arg1.$ === "DataLiteral"
            ? processRawSliceCode(arg1)
            : decompiledCode(processInstructions(ctx, arg1.instructions));

    return [first, second];
};

export const ifElseBitArgs = (
    ctx: Ctx,
    instr: $ast.Instruction,
): [number, Code, Code] => {
    const [arg0raw, arg1raw, arg2raw] = instr.args;
    if (!arg0raw || !arg1raw || !arg2raw) {
        throw new Error(`Expected 3 arguments`);
    }

    const arg0 = arg0raw.expression;
    if (arg0.$ !== "IntegerLiteral") {
        throw new Error(`Expected integer literal argument 1, got ${arg0.$}`);
    }

    const arg1 = arg1raw.expression;
    if (arg1.$ !== "Code" && arg1.$ !== "DataLiteral") {
        throw new Error(
            `Expected code or data literal argument 2, got ${arg1.$}`,
        );
    }

    const arg2 = arg2raw.expression;
    if (arg2.$ !== "Code" && arg2.$ !== "DataLiteral") {
        throw new Error(
            `Expected code or data literal argument 3, got ${arg2.$}`,
        );
    }

    const second: Code =
        arg1.$ === "DataLiteral"
            ? processRawSliceCode(arg1)
            : decompiledCode(processInstructions(ctx, arg1.instructions));
    const third =
        arg2.$ === "DataLiteral"
            ? processRawSliceCode(arg2)
            : decompiledCode(processInstructions(ctx, arg2.instructions));

    return [parseNumber(arg0), second, third];
};

export const ifBitArgs = (
    ctx: Ctx,
    instr: $ast.Instruction,
): [number, Code] => {
    const [arg0raw, arg1raw] = instr.args;
    if (!arg0raw || !arg1raw) {
        throw new Error(`Expected 2 arguments`);
    }

    const arg = arg0raw.expression;
    if (arg.$ !== "IntegerLiteral") {
        throw new Error(`Expected integer literal argument 1, got ${arg.$}`);
    }

    const arg1 = arg1raw.expression;
    if (arg1.$ !== "Code" && arg1.$ !== "DataLiteral") {
        throw new Error(
            `Expected code or data literal argument 2, got ${arg1.$}`,
        );
    }

    const second: Code =
        arg1.$ === "DataLiteral"
            ? processRawSliceCode(arg1)
            : decompiledCode(processInstructions(ctx, arg1.instructions));

    return [parseNumber(arg), second];
};

export const sliceArg = (instr: $ast.Instruction) => {
    const [arg0raw] = instr.args;
    if (!arg0raw) {
        throw new Error(`Expected 1 argument`);
    }

    const arg = arg0raw.expression;
    if (arg.$ !== "DataLiteral") {
        throw new Error(
            `Expected hex, bin or boc literal argument 1, got ${arg.$}`,
        );
    }

    return parseDataLiteral(arg);
};

export const debugstrArg = (instr: $ast.Instruction) => {
    const [arg0raw] = instr.args;
    if (!arg0raw) {
        throw new Error(`Expected 1 argument`);
    }

    const arg = arg0raw.expression;
    if (arg.$ !== "DataLiteral") {
        throw new Error(
            `Expected hex, bin or string literal argument 1, got ${arg.$}`,
        );
    }

    return parseDataLiteral(arg);
};

export const dictpushArg = (
    ctx: Ctx,
    instr: $ast.Instruction,
): [number, Dict] => {
    const [arg0raw, arg1raw] = instr.args;
    if (!arg0raw || !arg1raw) {
        throw new Error(`Expected 2 arguments`);
    }

    const arg0 = arg0raw.expression;
    if (arg0.$ !== "IntegerLiteral") {
        throw new Error(`Expected integer argument 1, got ${arg0.$}`);
    }

    const keyLength = parseNumber(arg0);

    const arg1 = arg1raw.expression;

    if (arg1.$ === "DataLiteral") {
        const slice = parseDataLiteral(arg1);
        return [keyLength, i.util.rawDict(slice)];
    }

    if (arg1.$ !== "Dictionary") {
        throw new Error(`Expected dictionary argument 2, got ${arg1.$}`);
    }

    const methods = arg1.entries.map((entry): DecompiledMethod => {
        const id = entry.id;
        const code = entry.code;

        return {
            $: "DecompiledMethod",
            id: parseNumber(id),
            instructions: processInstructions(ctx, code.instructions),
        };
    });

    return [keyLength, i.util.decompiledDict(methods)];
};

const parseDataLiteral = (literal: $ast.DataLiteral): Slice => {
    const arg = literal.value;

    if (arg.$ === "StringLiteral") {
        return beginCell().storeBuffer(Buffer.from(arg.value)).asSlice();
    }

    if (arg.$ === "HexLiteral") {
        return hex(arg.content);
    }

    if (arg.$ === "BocLiteral") {
        return boc(arg.content);
    }

    return bin(arg.content);
};

const parseStackElement = (arg: $ast.StackElement) =>
    Number.parseInt(arg.value.slice(1));

const parseNumber = (literal: $ast.IntegerLiteral) => {
    const val = Number.parseInt(literal.value.digits);
    if (literal.op === "-") {
        return -val;
    }
    return val;
};

const processRawSliceCode = (literal: $ast.DataLiteral): Code => {
    const slice = parseDataLiteral(literal);
    try {
        const decompiled = i.decompileCell(slice.asCell());
        return code(decompiled);
    } catch {
        return rawCode(slice);
    }
};
