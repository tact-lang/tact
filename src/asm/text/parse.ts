import * as $ from "@tonstudio/parser-runtime";
import * as G from "@/asm/text/grammar";
import type { Instr } from "@/asm/runtime";
import type { Ctx } from "@/asm/text/util";
import { createLoc, ParseError, processInstructions } from "@/asm/text/util";

export type ParseResult = ParseSuccess | ParseFailure;

export type ParseSuccess = {
    readonly $: "ParseSuccess";
    readonly instructions: Instr[];
};

export type ParseFailure = {
    readonly $: "ParseFailure";
    readonly error: ParseError;
};

const success = (instructions: Instr[]): ParseSuccess => ({
    $: "ParseSuccess",
    instructions,
});
const failure = (error: ParseError): ParseFailure => ({
    $: "ParseFailure",
    error,
});

export function parse(filepath: string, code: string): ParseResult {
    const lines = code.split("\n");
    const ctx: Ctx = { lines, filepath };

    const res = $.parse({
        grammar: G.File,
        space: G.space,
        text: code,
    });

    if (res.$ !== "success") {
        const { expected, position } = res.error;

        const loc = createLoc(ctx, { $: "empty", at: position });
        return failure(
            new ParseError(loc, `Expected ${getExpectedText(expected)}`),
        );
    }

    try {
        return success(processInstructions(ctx, res.value.instructions));
    } catch (error) {
        if (error instanceof ParseError) {
            return failure(error);
        }

        throw error;
    }
}

const getExpectedText = (expected: ReadonlySet<string>) => {
    const result: string[] = [];
    const failures = [...expected].sort();
    for (const [idx, failure] of failures.entries()) {
        if (idx > 0) {
            if (idx === failures.length - 1) {
                result.push(failures.length > 2 ? ", or " : " or ");
            } else {
                result.push(", ");
            }
        }
        result.push(failure);
    }
    return result.join("");
};
