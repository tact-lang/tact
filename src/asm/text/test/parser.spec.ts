import { decompileCell } from "@/asm/runtime";
import { print } from "@/asm/text/printer";
import { readFileSync } from "node:fs";
import { parse } from "@/asm/text/parse";
import { boc } from "@/asm/runtime/util";

describe("assembly-parser", () => {
    it("should parse simple assembly", () => {
        const code = `
            PUSHINT 10
            PUSHINT 5
            ADD
        `;
        const res = parse("test.asm", code);
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error");
        }

        expect(print(res.instructions)).toMatchSnapshot();
    });

    it("should parse assembly with raw pushref", () => {
        const code = `
            PUSHREF x{71}
        `;
        const res = parse("test.asm", code);
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error");
        }

        expect(print(res.instructions)).toMatchSnapshot();
    });

    it("should parse assembly with invalid raw pushref", () => {
        const code = `
            PUSHREF x{22221}
        `;
        const res = parse("test.asm", code);
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error");
        }

        expect(print(res.instructions)).toMatchSnapshot();
    });

    it("should parse and print assembly", () => {
        const instructions = decompileCell(
            boc(
                readFileSync(
                    `${__dirname}/testdata/jetton_minter_discoverable_JettonMinter.boc`,
                ).toString("hex"),
            ).asCell(),
        );
        const assembly = print(instructions);

        const res = parse("test.asm", assembly);
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error");
        }

        const assembly2 = print(res.instructions);

        expect(assembly2).toEqual(assembly);
    });

    it("should not parse assembly with error", () => {
        const code = `
            PUSHINT 10 ,
            PUSHINT 5
            ADD
        `;
        const res = parse("test.asm", code);
        if (res.$ === "ParseSuccess") {
            throw new Error("unexpected parser success");
        }

        expect(res.error.toString()).toMatchSnapshot();
    });

    it("should give an error for malformed assembly", () => {
        const code = `
            PUSHINT // no arg
            PUSHINT 5
            ADD
        `;
        const res = parse("test.asm", code);
        if (res.$ === "ParseSuccess") {
            throw new Error("unexpected parser success");
        }

        expect(res.error.toString()).toMatchSnapshot();
    });
});
