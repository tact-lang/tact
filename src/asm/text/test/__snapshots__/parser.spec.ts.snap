// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assembly-parser should give an error for malformed assembly 1`] = `"ParseError: Expected 1 argument at test.asm:1"`;

exports[`assembly-parser should not parse assembly with error 1`] = `"ParseError: Expected "-", "0", "[", "\\"", "boc{", "b{", "c", "embed", "exotic", "ref", "s", "x{", "{", digit, or identifier at test.asm:1"`;

exports[`assembly-parser should parse assembly with invalid raw pushref 1`] = `
"PUSHREF {
    embed x{22221}
}
"
`;

exports[`assembly-parser should parse assembly with raw pushref 1`] = `
"PUSHREF {
    PUSHINT 1
}
"
`;

exports[`assembly-parser should parse simple assembly 1`] = `
"PUSHINT 10
PUSHINT 5
ADD
"
`;
